### PR TITLE
Introduce types to manage the state of the build process and daemon

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

These are analogs to `BuildSessionState`, `BuildState`, etc.

Also:
- introduce a new project to contain the types related to build state management. The intention is that some types from `core` and `launcher` will move into this new project, but in this PR it contains only the new type.
- apply validation to global scope services in the daemon and tooling API
- properly clean up global services when the daemon exits.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/cli/BuildActionsFactory.java [hunk 1]
```diff
@@ -243,21 +239,17 @@ private Runnable runBuildInSingleUseDaemon(StartParameterInternal startParameter
...
         DaemonClient client = clientServices.get(DaemonClient.class);
         return runBuildAndCloseServices(startParameter, daemonParameters, client, clientSharedServices, clientServices);
     }
 
-    private ServiceRegistry createGlobalClientServices(boolean usingDaemon) {
+    private ServiceRegistry createGlobalClientServices() {
         ServiceRegistryBuilder builder = ServiceRegistryBuilder.builder()
             .displayName("Daemon client global services")
             .parent(NativeServices.getInstance());
-        if (usingDaemon) {
```
Comment (adammurdoch) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/cli/BuildActionsFactory.java:256]: This conditional was not necessary. The "basicServices" could be used in all cases.

File: platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/BuildProcessState.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.gradle.internal.service.scopes.GlobalScopeServices;
+import org.gradle.internal.service.scopes.GradleUserHomeScopeServiceRegistry;
+import org.gradle.internal.service.scopes.Scope;
+
+import java.io.Closeable;
+
+/**
+ * Encapsulates the state of a build process, such as the Gradle daemon. An instance is created for each process that runs a build.
+ */
+public class BuildProcessState implements Closeable {
```
Comment (adammurdoch) [inline platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/BuildProcessState.java:33]: This is the main change, to extract this logic from the various places that need to set up a build process into one place.

File: platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonProcessState.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import org.gradle.launcher.daemon.configuration.DaemonServerConfiguration;
+import org.gradle.launcher.daemon.registry.DaemonRegistryServices;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Encapsulates the state of the daemon process.
+ */
+public class DaemonProcessState implements Closeable {
```
Comment (adammurdoch) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonProcessState.java:35]: Use a type to manage the daemon process state. There's still some logic in the daemon's main()` method that could move into this type, but for this PR, simply wrap the daemon's `BuildProcessState` and service lifecycling logic.

File: platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/BuildProcessState.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    public BuildProcessState(
+        final boolean longLiving,
+        AgentStatus agentStatus,
+        ClassPath additionalModuleClassPath,
+        ServiceRegistry... parents
+    ) {
+        ServiceRegistryBuilder builder = ServiceRegistryBuilder.builder()
+            .scopeStrictly(Scope.Global.class)
+            .displayName("Global services")
```
Comment (big-guy) [inline platforms/core-runtime/build-state/src/main/java/org/gradle/internal/buildprocess/BuildProcessState.java:44]: 💭 Is there any value in this name being more specific to which build process we're in?

**Review by big-guy — APPROVED** [review]

**Pull request merged**